internal/operator-controller/rukpak/convert/installer_rbac.go

```diff
@@ -0,0 +1,166 @@
+	return rbacv1.PolicyRule{
+		APIGroups:     []string{"olm.operatorframework.io"},
+		Resources:     []string{"clusterextensions/finalizers"},
+		Verbs:         []string{"update"},
+		ResourceNames: []string{clusterExtensionName},
+	}
+}
+
+// GenerateResourceManagerRolePerms generates role permissions to manage objs resources in their
+// namespaces. The permissions also include any permissions defined in any Roles in objs within the namespace, allowing
+// the holder to also assign the RBAC therein to another service account.
+// Note: currently assumes objs have been created by convert.Convert.
+// The returned Roles will not have set .metadata.name
+func GenerateResourceManagerRolePerms(objs []client.Object) map[string][]rbacv1.PolicyRule {
+	out := map[string][]rbacv1.PolicyRule{}
+	namespaceScopedObjs := filter.Filter(objs, isNamespaceScopedResource)
+	for _, obj := range namespaceScopedObjs {
+		namespace := obj.GetNamespace()
+		if _, ok := out[namespace]; !ok {
+			objsInNamespace := filter.Filter(namespaceScopedObjs, isInNamespace(namespace))
+			out[namespace] = slices.Concat(
+				// namespace scoped resource creation and management rules
+				generatePolicyRules(objsInNamespace),
+				// controller rbac scope
+				collectRBACResourcePolicyRules(filter.Filter(objsInNamespace, filter.And(isOfKind("Role"), isGeneratedResource))),
+			)
+		}
+	}
+	return out
+}
+
+func generatePolicyRules(objs []client.Object) []rbacv1.PolicyRule {
+	return slices.Concat(
+		mapToSlice(slicesutil.GroupBy(objs, groupKind), func(gk schema.GroupKind, resources []client.Object) []rbacv1.PolicyRule {
+			return []rbacv1.PolicyRule{
+				newPolicyRule(gk, unnamedResourceVerbs),
+				newPolicyRule(gk, namedResourceVerbs, slicesutil.Map(resources, toResourceName)...),
+			}
+		})...,
+	)
+}
+
+func collectRBACResourcePolicyRules(objs []client.Object) []rbacv1.PolicyRule {
+	return slices.Concat(slicesutil.Map(objs, func(obj client.Object) []rbacv1.PolicyRule {
+		if cr, ok := obj.(*rbacv1.ClusterRole); ok {
+			return cr.Rules
+		} else if r, ok := obj.(*rbacv1.Role); ok {
+			return r.Rules
+		} else {
+			panic(fmt.Sprintf("unexpected type %T", obj))
+		}
+	})...)
+}
+
+func newPolicyRule(groupKind schema.GroupKind, verbs []string, resourceNames ...string) rbacv1.PolicyRule {
+	return rbacv1.PolicyRule{
+		APIGroups:     []string{groupKind.Group},
+		Resources:     []string{fmt.Sprintf("%ss", strings.ToLower(groupKind.Kind))},
+		Verbs:         verbs,
+		ResourceNames: resourceNames,
+	}
+}
+
+func mapToSlice[K comparable, V any, R any](m map[K]V, fn func(k K, v V) R) []R {
+	out := make([]R, 0, len(m))
+	for k, v := range m {
+		out = append(out, fn(k, v))
+	}
+	return out
+}
+
+func isClusterScopedResource(o client.Object) bool {
+	return slices.Contains(clusterScopedResources, o.GetObjectKind().GroupVersionKind().Kind)
+}
+
+func isNamespaceScopedResource(o client.Object) bool {
```

Comment:
since we expect the rendered manifests as input, I'm wondering if we could just check whether `.metadata.namespace` is set as the discriminant....